### PR TITLE
MONGOID-4817 use #discriminator_key function instead of _type

### PR DESCRIPTION
**Pull request by Neilshweky**
So far I have changed criteria.rb to no longer hard-code _type.

### CODE REVIEW

**Comment by Neilshweky** [comment]
It seems like the `selector_with_type_selection` function is never used... Not sure why it's still there, but it's the only place that `type_selectable?` is used... I changed it anyway to use `#discriminator_key` instead of `_type` and I guess it doesn't hurt to leave in extra functions, I just find it weird that they're there (and `api private`).

File: lib/mongoid/criteria.rb [hunk 1]
```diff
@@ -555,8 +555,7 @@ def merge_type_selection
     # @since 3.0.3
     def type_selectable?
       klass.hereditary? &&
-        !selector.keys.include?("_type") &&
-        !selector.keys.include?(:_type)
+        !selector.keys.include?(self.discriminator_key)
```
Comment (p-mongo) [inline lib/mongoid/criteria.rb:558]: This still needs to account for string and symbol keys.

**Comment by p-mongo** [comment]
`selector_with_type_selection` is currently not used by anything. However, `type_selectable?` *is* used by Mongoid.

File: lib/mongoid/criteria.rb [hunk 2]
```diff
@@ -292,7 +292,7 @@ def only(*args)
         args.unshift(:_id)
       end
       if klass.hereditary?
-        super(*args.push(:_type))
+        super(*args.push(klass.discriminator_key))
```
Comment (p-mongo) [inline lib/mongoid/criteria.rb:295]: This was changed from symbol to string.

File: lib/mongoid/factory.rb [hunk 3]
```diff
@@ -7,8 +7,6 @@ module Mongoid
   module Factory
     extend self
 
-    TYPE = "_type".freeze
```
Comment (p-mongo) [inline lib/mongoid/factory.rb:10]: This constant needs to stay here with `@deprecated` tag.

File: spec/mongoid/document_spec.rb [hunk 4]
```diff
@@ -881,27 +881,60 @@ class Manager < Person
...
+          it "sets the class type" do
+            expect(person._type).to eq("Person")
+          end
 
-        it "sets the class type" do
-          expect(person._type).to eq("Person")
+          it "raises an error when inappropriate class is provided" do
+            expect {
+              manager.becomes(String)
+            }.to raise_error(ArgumentError)
```
Comment (p-mongo) [inline spec/mongoid/document_spec.rb:904]: Can you add a regexp match on the exception message here please?

File: spec/mongoid/inspectable_spec.rb [hunk 5]
```diff
@@ -32,6 +32,33 @@
       it "displays field aliases" do
         expect(inspected).to include("t(test):")
       end
+
+      it "displays the default discriminator key" do 
+        expect(inspected).to include("_type: \"Person\"")
```
Comment (p-mongo) [inline spec/mongoid/inspectable_spec.rb:37]: Can you replace the string with e.g. `%q,_type: "Person",` please.

File: lib/mongoid/criteria.rb (hunk 2 above)
Comment (Neilshweky) [reply lib/mongoid/criteria.rb:295]: do you want me to put a `.to_sym` here?

File: lib/mongoid/criteria.rb (hunk 1 above)
Comment (Neilshweky) [reply lib/mongoid/criteria.rb]: I fixed this, but I don't understand why it was needed. `self.discriminator_key` is going to equal whatever the user set the discriminator key, whether it's a symbol or a string, no?

**Pull request closed without merging**